nowcasting_dataset/data_sources/datasource_output.py

```diff
@@ -0,0 +1,155 @@
+""" General Data Source output pydantic class. """
+from pydantic import BaseModel, Field
+import pandas as pd
+import xarray as xr
+import numpy as np
+from typing import List
+import logging
+
+from nowcasting_dataset.utils import to_numpy
+
+logger = logging.getLogger(__name__)
+
+
+class DataSourceOutput(BaseModel):
+    """General Data Source output pydantic class.
+
+    Data source output classes should build of this class
+    """
+
+    class Config:
+        """ Allowed classed i.e. tensor.Tensor"""
+
+        # TODO maybe there is a better way to do this
+        arbitrary_types_allowed = True
+
+    batch_size: int = Field(
+        0,
+        ge=0,
+        description="The size of this batch. If the batch size is 0, "
+        "then this item stores one data item",
+    )
+
+    def to_numpy(self):
+        """Change to numpy"""
+        for k, v in self.dict().items():
+            self.__setattr__(k, to_numpy(v))
+
+    def to_xr_data_array(self):
+        """ Change to xr DataArray"""
+        pass
+
+    @staticmethod
+    def join(data):
+        """ Join a list of data source items to a batch. """
+        _ = [d.to_numpy() for d in data]
+
+        # use the first item in the list, and then update each item
+        batch = data[0]
+        for k, v in batch.dict().items():
+
+            # set batch size to the list of the items
+            if k == "batch_size":
+                batch.batch_size = len(data)
+            else:
+
+                # get list of one variable from the list of data items.
+                one_variable_list = [d.__getattribute__(k) for d in data]
+                batch.__setattr__(k, np.stack(one_variable_list, axis=0))
+
+        return batch
+
+    def split(self):
+        """
+        Split the datasource from a batch to a list of items
+
+        Returns: List of single data source items
+        """
+        c = self.__class__
+
+        items = []
+        for batch_idx in range(self.batch_size):
+            d = {k: v[batch_idx] for k, v in self.dict().items() if k != "batch_size"}
+            d["batch_size"] = 0
+            items.append(c(**d))
+
+        return items
+
+    def to_xr_dataset(self):
+        """ Make a xr dataset. Each data source needs to defined this """
+        raise NotImplementedError
+
+    def from_xr_dataset(self):
+        """ Load from xr dataset. Each data source needs to defined this """
+        raise NotImplementedError
+
+    def select_time_period(
```

Comment:
So, I could be wrong, but I think the main use-case for this sort of function is to limit the number of timesteps of each example.  For example, perhaps the pre-prepared batches on disk have a history length of 2 hours, and a forecast length of 4 hours.  But a particular model only wants an hour of history and an hour of forecast.

So, instead of asking the user to specify a `time_of_first_example`, `start_time`, and `end_time`; it may instead be more useful if the user just provides a new `history_minutes` and `forecast_minutes`, and the returned batch is cropped.

Our imaginary user who only wants an hour of history and an hour of forecast would just call `select_time_period(history_miniutes=60, forecast_minutes=60)` and they'd get back a cropped batch.

But, perhaps there's another use-case that I've forgotten about?!

Reply:
I think that makes sense, the only other use case I can think of is if the imaginary user wants to change the current time of the batch, and take the current time as both an hour in the past, as well as 2 hours in the future as well, to get more examples out of each batch.